lib/modules/manager/npm/extract/pnpm.ts

```diff
@@ -135,3 +142,124 @@ export async function detectPnpmWorkspaces(
     }
   }
 }
+
+export async function getPnpmLock(filePath: string): Promise<LockFile> {
+  // TODO #7154
+  const pnpmLockRaw = (await readLocalFile(filePath, 'utf8'))!;
+  try {
+    const lockParsed = load(pnpmLockRaw);
+    if (!isPnpmLockfile(lockParsed)) {
+      throw new Error('Invalid or empty lockfile');
+    }
+    logger.trace({ lockParsed }, 'pnpm lockfile parsed');
+
+    // field lockfileVersion is type string in lockfileVersion = 6 and type number in < 6
+    const lockfileVersion: number = is.number(lockParsed.lockfileVersion)
+      ? lockParsed.lockfileVersion
+      : parseFloat(lockParsed.lockfileVersion);
+
+    const lockedVersions: Record<string, string> = {};
+    const packagePathRegex = regEx(
+      /^\/(?<packageName>.+)(?:@|\/)(?<version>[^/@]+)$/
+    ); // eg. "/<packageName>(@|/)<version>"
+
+    for (const packagePath of Object.keys(lockParsed.packages ?? {})) {
+      const result = packagePath.match(packagePathRegex);
+      if (!result?.groups) {
+        logger.debug(`Invalid package path ${packagePath}`);
+        continue;
+      }
+
+      const packageName = result.groups.packageName;
+      const version = result.groups.version;
+      logger.trace({
+        packagePath,
+        packageName,
+        version,
+      });
+      lockedVersions[packageName] = version;
+    }
+    return {
+      lockedVersions,
+      lockfileVersion,
+    };
+  } catch (err) {
+    logger.debug({ filePath, err }, 'Warning: Exception parsing pnpm lockfile');
+    return { lockedVersions: {} };
+  }
+}
+
+export function getConstraints(
+  lockfileVersion: number,
+  constraints?: string
+): string {
+  let newConstraints = constraints;
+
+  // find matching lockfileVersion
+  // fallback to version:5 if no match is found because `pnpm-lock.yaml` can only be generated using pnpm >= 5
+  const { lowerBound, upperBound, lowerConstraint, upperConstraint } =
+    lockToPnpmVersionMapping.find(
+      (m) => m.lockfileVersion === lockfileVersion
+    ) ?? {
+      lockfileVersion: 5.0,
+      lowerBound: '2.24.0',
+      upperBound: '3.5.0',
+      lowerConstraint: '>=3',
+      upperConstraint: '<3.5.0',
+    };
+
+  if (newConstraints) {
+    // if constraint satisfies versions lower than min allowed version
+    if (semver.satisfies(lowerBound, newConstraints)) {
+      newConstraints += ` ${lowerConstraint}`;
+    }
+
+    // if constraint satisfies versions higher than max allowed version
+    if (
+      upperBound &&
+      upperConstraint &&
+      semver.satisfies(upperBound, newConstraints)
+    ) {
+      newConstraints += ` ${upperConstraint}`;
+    }
+  } else {
+    newConstraints = `${lowerConstraint}${
+      upperConstraint ? ` ${upperConstraint}` : ''
+    }`;
+  }
+
+  return newConstraints;
+}
+
+// lockfileVersion, maxNotAllowedVersion, minAllowedVersion
+const lockToPnpmVersionMapping = [
```

Comment:
Please describe this in more depth, and include what to do about future lock file versions